### ISSUE TITLE
Wrong AWS volume can be mounted

### ISSUE THREAD

**Issue by atombender** [kubernetes/kubernetes#29324]
Background: K8s often has problems attaching/detaching AWS volumes: If I do a `kubectl replace` on a deployment, I frequently observe K8s not correctly detaching the volume, and the new deployment will be forever stuck in a crash loop because it times out getting the new volume. Seems like a race condition. The steps I always do when this happens is to delete the deployment, manually force-detach via EC2 API, and manually `umount` the mounts, otherwise they never seem to be cleaned up.

So today I experienced this again, and I did the following:
1. Delete deployment.
2. Manually detach volume via EC2 API, since it was not being released.
3. **Manually `umount` wrong volume**. I accidentally unmounted the volume of a different pod because I didn't read the `mount` output correctly. Oops.
4. Created deployment.
5. Noticed my mistake.

What happened was that the new deployment was mounted with the wrong device:

``` shell
$ mount | grep xvdb
/dev/xvdba on /var/lib/kubelet/plugins/kubernetes.io/aws-ebs/mounts/vol-0e3bf8343563ee5aa type ext4 (rw,relatime,data=ordered)
/dev/xvdba on /var/lib/kubelet/pods/a8f54bf6-4ebd-11e6-89fc-065b459e106f/volumes/kubernetes.io~aws-ebs/data type ext4 (rw,relatime,data=ordered)
```

This is the right volume _ID_, but it has mounted the wrong device. According to EC2, the (right) volume is attached to `/dev/xvdbb`.

``` shell
$ aws ec2 describe-volumes --volume-ids vol-0e3bf8343563ee5aa
{
    "Volumes": [
        {
            "AvailabilityZone": "us-east-1e", 
            "Attachments": [
                {
                    "AttachTime": "2016-07-20T21:05:17.000Z", 
                    "InstanceId": "i-06d03923e9e5a6329", 
                    "VolumeId": "vol-0e3bf8343563ee5aa", 
                    "State": "attached", 
                    "DeleteOnTermination": false, 
                    "Device": "/dev/xvdbb"
                }
            ], 
            "Tags": [],
            "Encrypted": false, 
            "VolumeType": "gp2", 
            "VolumeId": "vol-0e3bf8343563ee5aa", 
            "State": "in-use", 
            "Iops": 100, 
            "SnapshotId": "", 
            "CreateTime": "2016-07-19T00:25:00.168Z", 
            "Size": 30
        }
    ]
}
```

`/dev/xvdba` is the volume I accidentally unmounted by mistake:

``` javascript
            // ...
            "Attachments": [
                {
                    "AttachTime": "2016-07-19T17:56:27.000Z", 
                    "InstanceId": "i-06d03923e9e5a6329", 
                    "VolumeId": "vol-0e0a13f1b3a4a6dbc", 
                    "State": "attached", 
                    "DeleteOnTermination": false, 
                    "Device": "/dev/xvdba"
                }
            ], 
```

So it's mounting the wrong volume into the pod.

Of course I was being naughty, but it seems to me that this should never happen, since EC2 has already allocated `xvdbb`?

[Here's](https://gist.github.com/atombender/a1c8199ed202c2ea0d6b4566db65c60b) the entire KCM output from this fiasco.

**Comment by sstarcher** [comment]
I may of just ran into the same problem without doing any funny work.  I have 2 separate pods that mount 2 separate aws ebs volumes.  Currently on the instance only 1 volume shows up as mounted, but in AWS EC2 both volumes are attached to the instance.  Both pods have the same data volume mounted inside.

Kubernetes 1.3.4
* Pod A has a persistent volume claim and mounts ebs volume A
* Pod B has a direct awsElasticBlockStore statement in the pod for ebs volume B
* Volume A and B are both attached to the EC2 instance
* Volume A is properly mounted on the host
* Pod A and Pod B both have volume A mounted inside of them.  This is incorrect for Pod B.
* Deleting Pod A unattaches the ebs volume A and ebs volume B was mounted

**Comment by jingxu97** [comment]
@atombender, @sstarcher Thank you for posting the issues. Could you please share the kubelet log if it is still available? Thanks!

Jing

**Comment by atombender** [comment]
@jingxu97 https://gist.github.com/atombender/a810ca55fd1e149f20d2a7a6f7a45934

**Comment by jingxu97** [comment]
@atombender , thanks for the log. Could you also please share your deployment spec with us? Thanks!

**Comment by atombender** [comment]
I don't have it right here, but I basically just use an AWS volume in the pod declaration (_not_ a persistent vol + claim) and a single volume mount.

I will send you the actual file later when I am back before a screen.

**Comment by jingxu97** [comment]
@atombender thanks for your response. Also the log file you posted only has some error messages. Do you still have the kubelet full log on the instance? Thanks!

**Comment by atombender** [comment]
That is the full log. (I have kept the logs all the way from when I started using K8s.) I believe I extracted the right time range. 

FWIW, I don't think K8s logged much in terms of errors (though you will see some in the KCM log). In fact, that's kind of the point -- it attached the wrong thing, without erroring.

**Comment by jingxu97** [comment]
@sstarcher  I followed your steps but with gce disks, and it seems mounting correctly. Could you please share the kubelet logs so that I can do some debugging? Thanks!

**Comment by jingxu97** [comment]
@atombender From KCM log, I only see attach and detach once. But it seems that you created two deployments, and it should have two attach log. From kubelet log, they are all error messages (unable to mount, unable to cleanup), but there is no information log.

**Comment by atombender** [comment]
I _may_ have done the other deploy the day before. I will get you logs going farther back in time.

**Comment by sstarcher** [comment]
@jingxu97 I don't have the same exact setup anymore.  I'll attempt to reproduce on monday if possible.

**Comment by jingxu97** [comment]
@justinsb I was trying to reproduce the error @sstarcher described for AWS volume. I tried the same steps on gce disks and it is working fine. But I don't have the AWS volumes to test. If you have the setup, could you please try it on your environment when you have time.  The steps @sstarcher described as follows. Thank you very much!

Kubernetes 1.3.4

- Pod A has a persistent volume claim and mounts ebs volume A
- Pod B has a direct awsElasticBlockStore statement in the pod for ebs volume B
- Volume A and B are both attached to the EC2 instance
- Volume A is properly mounted on the host
- Pod A and Pod B both have volume A mounted inside of them. This is incorrect for Pod B.
- Deleting Pod A unattaches the ebs volume A and ebs volume B was mounted

**Comment by justinsb** [comment]
I think I know what is happening here...

* A user reported that this was happening also with just awsElasticBlockStore volumes.  They provided me an extract of their logs.  (Thanks Fairfax!)
* In the logs, we saw ~5 volumes being assigned to a single instance at the same time, and all 5 of them were allocated the _same_ device (within a very small time interval).
* This is supposed to be impossible - we keep an in-memory map of mounts in-flight precisely to avoid this, so that we can keep track of devices before they hit the AWS API (because we have to choose a device for AWS to mount the volume, but I didn't particularly want to hold a lock while making the attachment)
* BUT... we keep this map for the local instance, because mounts happen locally, but not on the master because otherwise we'd have to worry about how to garbage collect it.
* but of course in 1.3 the previous doesn't hold - mounts now happen on the master :-(

Looking at simple ways to fix this.  Probably making the map of mounts in-flight just be global state.  It _shouldn't_ leak, but even if it does under rare circumstances, it will be better than this bug.

**Comment by justinsb** [comment]
Just put up an experimental fix - haven't tested it even compiles yet.  But thought I would get the ball rolling on how I think we should fix this!

**Comment by mrook** [comment]
We've also had this happen when a number of machines (that had a substantial number of volumes attached to them) were recycled in a short period of time. Looking forward to the fix!

**Comment by svanderbijl** [comment]
@justinsb even with a patched kube-controller-manager with your fixes we can consistently reproduce volumes being mounted/assigned to the wrong pods. Your updates do seem to solve volumes being assigned to multiple pods but the ordering is wrong (volume 1 belong to pod-1 now assigned to pod-2 and vice-versa).

Is the kube-controller-manager the only component affected by this or should we also replace e.g. kubelet? 

Thanks for your help!

**Comment by svanderbijl** [comment]
For anyone running into this problem. We've disabled the new controller attach/detach logic for now by passing the flag --enable-controller-attach-detach=false to kubelet. Volumes are now mounted correctly.

**Comment by jingxu97** [comment]
Hi @mrook @svanderbijl Thank you for reporting the issue with us. In order to have a right fix for the problem, we need to reproduce the error. Could you please share your detailed information about your cluster setup and pod deployment specs etc. so that we can try to run the tests? Thank you!

**Comment by jingxu97** [comment]
@sstarcher Could you please let us know whether you can reproduce the issue you reported in this thread? Thank you!

**Comment by sstarcher** [comment]
@jingxu97 we have not had the issue in the past 2 weeks.

**Comment by saad-ali** [comment]
> @justinsb even with a patched kube-controller-manager with your fixes we can consistently reproduce volumes being mounted/assigned to the wrong pods. Your updates do seem to solve volumes being assigned to multiple pods but the ordering is wrong (volume 1 belonging to pod-1 now assigned to pod-2 and vice-versa).
>
> Is the kube-controller-manager the only component affected by this or should we also replace e.g. kubelet? 

For AWS, Kubelet just uses the device path returned by `Attach(...)` (the master passes it to the node, if the node is not doing its own attaches), so you shouldn't need to patch kubelet. @justinsb any thoughts? Is the AWS AttachDisk code making other assumptions that are being violated when not executing on node that could cause it to return the wrong path on `AttachDisk(...)`?

@mrook, @svanderbijl: It sounds like you have a consistent repro. We've had trouble reproing this issue. As @jingxu97 mentioned, if you could share the repro steps with us that would make fixing this a whole lot easier.

**Comment by jingxu97** [comment]
@justinsb I checked some code in attacher, will this cause any problem in returning the device path?
https://github.com/kubernetes/kubernetes/blob/c5ab95cd792c24e4d0c2fc561e79856ac8236a5b/pkg/volume/aws_ebs/attacher.go#L105

**Comment by justinsb** [comment]
@jingxu97 I'm not seeing the problem, but I could well be missing something.

Also, from the very beginning I wanted this to share GCE's code (rather than being a copy-paste), because they inevitably drift.  This is supposed to be a copy of https://github.com/kubernetes/kubernetes/blob/c5ab95cd792c24e4d0c2fc561e79856ac8236a5b/pkg/volume/gce_pd/attacher.go#L126, but it looks like they have drifted..

**Comment by saad-ali** [comment]
I don't see anything wrong with that particular bit of code either @jingxu97.

@justinsb the drift in this case is necessary because GCE is able to deterministically compute the expected device path for a given GCE PD, whereas AWS EBS device path's are non-deterministic, returned by the attach operation, and must be passed to kubelet/mount operations.

**Comment by svanderbijl** [comment]
@justinsb @saad-ali we configured a Kubernetes testcluster using our own Terraform setup. Controller + two workers, both in different AZ. We deployed 20 petsets to the cluster (each claiming a 1Gb volume). All the petsets do is write their container-id / hostname to the volume.  

We then destroy both worker nodes and wait for new nodes to be created and the pods to be deployed. When completed we verify pods got their own volume by verifying the contents of the testfile on the volume matches the hostname.

Scripts and templates we used can be found in https://github.com/fourscouts/filetestding.

**Comment by saad-ali** [comment]
@svanderbijl Thank you very much for sharing repro steps. How do you destroy the nodes? Delete the VM and then the k8s API object, vice versa?

@justinsb How would https://github.com/kubernetes/kubernetes/pull/31090 react to a node being deleted? It shouldn't matter right? Since the attach and detach are independent per node.

**Comment by svanderbijl** [comment]
We used the AWS console to terminate the instances. K8S is configured with the AWS cloud provider and detects no longer be present and cleans up.

**Comment by mrook** [comment]
An alternative strategy is to cordon the two workers first, deploy the petsets, then uncordon both workers.

**Comment by justinsb** [comment]
To confirm: this is with 1.3.5, and you do see the error with your reproduction?  I'm going to try to reproduce it this morning with your petsets.

**Comment by justinsb** [comment]
@saad-ali in general the AWS provider should cope with nodes being deleted (with and without the patch - it is a very normal operation on AWS), though I hit #31088 when I tried it before

**Comment by mrook** [comment]
@justinsb yes, with vanilla 1.3.5 and the petset example we can consistently reproduce wrong volume assignments

**Comment by justinsb** [comment]
Confirmed with 1.3.6 on AWS:

```
Uit file:
filetestding-4-1
Hostname:
filetestding-16-1
Sleeping...
```

I used the cordon approach.  Looking at logs now.

My notes:

```
./generate.sh
find *.yaml | xargs -I {} kubectl create -f {}

sleep 120

PODS=`kubectl get pods -o name | cut -f 2 -d '/'`

for pod in $PODS; do echo "*******"; echo $pod; kubectl logs ${pod}; done


kubectl cordon ip-172-20-78-99.ec2.internal 
kubectl cordon ip-172-20-78-100.ec2.internal   

for pod in $PODS; do kubectl delete pod ${pod}; done


kubectl uncordon ip-172-20-78-99.ec2.internal 
kubectl uncordon ip-172-20-78-100.ec2.internal   


sleep 120

PODS=`kubectl get pods -o name | cut -f 2 -d '/'`

for pod in $PODS; do echo "*******"; echo $pod; kubectl logs ${pod}; done

```

**Comment by justinsb** [comment]
Here is what I have found so far:

* We hit AWS rate limits with this many volumes (which means requests taking longer because we insert delays).  So we are likely going into a rarer execution path somewhere. 
* The volume vol-c85ad849 was initially attached to "bs", then the second time to "bm"
* The volume was rescheduled onto the same node
* The interval between the Detach completing and the Attach starting was very short  (20ms).  Though the attach itself did not complete until minutes later.
* The node shows the correct mountpoint: `- devicePath: /dev/xvdbm` (after everything has completed)
* EC2 has the volume mounted on bm
* *BUT* The node has the volume mounted on the old mountpoint: `/dev/xvdbs on /var/lib/kubelet/plugins/kubernetes.io/aws-ebs/mounts/aws/us-east-1b/vol-c85ad849 type ext4 (rw,relatime,data=ordered)`


kubelet logs do show it mounting the volume on the first mountpoint (bs), formatting it, releasing it, then seeing the volume as absent repeatedly:

```
Sep 01 13:01:19 ip-172-20-78-100 kubelet[738]: E0901 13:01:19.416553     738 nestedpendingoperations.go:233] Operation for "\"kubernetes.io/aws-ebs/aws://us-east-1b/vol-c85ad849\"" failed. No retries permitted until 2016-09-01 13:01:20.416526582 +0000 UTC (durationBeforeRetry 1s). Error: Volume "kubernetes.io/aws-ebs/aws://us-east-1b/vol-c85ad849" (spec.Name: "pvc-47ad28c0-7043-11e6-849c-06cdf0c19b2d") pod "2bf9e9c1-7044-11e6-849c-06cdf0c19b2d" (UID: "2bf9e9c1-7044-11e6-849c-06cdf0c19b2d") has not yet been added to the list of VolumesInUse in the node's volume status.  
Sep 01 13:01:20 ip-172-20-78-100 kubelet[738]: I0901 13:01:20.420340     738 reconciler.go:180] VerifyControllerAttachedVolume operation started for volume "kubernetes.io/aws-ebs/aws://us-east-1b/vol-c85ad849" (spec.Name: "pvc-47ad28c0-7043-11e6-849c-06cdf0c19b2d") pod "2bf9e9c1-7044-11e6-849c-06cdf0c19b2d" (UID: "2bf9e9c1-7044-11e6-849c-06cdf0c19b2d")
Sep 01 13:01:20 ip-172-20-78-100 kubelet[738]: E0901 13:01:20.420469     738 nestedpendingoperations.go:233] Operation for "\"kubernetes.io/aws-ebs/aws://us-east-1b/vol-c85ad849\"" failed. No retries permitted until 2016-09-01 13:01:22.420447333 +0000 UTC (durationBeforeRetry 2s). Error: Volume "kubernetes.io/aws-ebs/aws://us-east-1b/vol-c85ad849" (spec.Name: "pvc-47ad28c0-7043-11e6-849c-06cdf0c19b2d") pod "2bf9e9c1-7044-11e6-849c-06cdf0c19b2d" (UID: "2bf9e9c1-7044-11e6-849c-06cdf0c19b2d") has not yet been added to the list of VolumesInUse in the node's volume status.  
Sep 01 13:01:22 ip-172-20-78-100 kubelet[738]: I0901 13:01:22.427419     738 reconciler.go:180] VerifyControllerAttachedVolume operation started for volume "kubernetes.io/aws-ebs/aws://us-east-1b/vol-c85ad849" (spec.Name: "pvc-47ad28c0-7043-11e6-849c-06cdf0c19b2d") pod "2bf9e9c1-7044-11e6-849c-06cdf0c19b2d" (UID: "2bf9e9c1-7044-11e6-849c-06cdf0c19b2d")
Sep 01 13:01:22 ip-172-20-78-100 kubelet[738]: E0901 13:01:22.427590     738 nestedpendingoperations.go:233] Operation for "\"kubernetes.io/aws-ebs/aws://us-east-1b/vol-c85ad849\"" failed. No retries permitted until 2016-09-01 13:01:26.427568736 +0000 UTC (durationBeforeRetry 4s). Error: Volume "kubernetes.io/aws-ebs/aws://us-east-1b/vol-c85ad849" (spec.Name: "pvc-47ad28c0-7043-11e6-849c-06cdf0c19b2d") pod "2bf9e9c1-7044-11e6-849c-06cdf0c19b2d" (UID: "2bf9e9c1-7044-11e6-849c-06cdf0c19b2d") has not yet been added to the list of VolumesInUse in the node's volume status.  
```

and then finally succeeding but *with the old mountpoint* (bs):

```
Sep 01 13:01:26 ip-172-20-78-100 kubelet[738]: I0901 13:01:26.441050     738 reconciler.go:180] VerifyControllerAttachedVolume operation started for volume "kubernetes.io/aws-ebs/aws://us-east-1b/vol-c85ad849" (spec.Name: "pvc-47ad28c0-7043-11e6-849c-06cdf0c19b2d") pod "2bf9e9c1-7044-11e6-849c-06cdf0c19b2d" (UID: "2bf9e9c1-7044-11e6-849c-06cdf0c19b2d")
Sep 01 13:01:26 ip-172-20-78-100 kubelet[738]: I0901 13:01:26.443278     738 operation_executor.go:980] Controller successfully attached volume "kubernetes.io/aws-ebs/aws://us-east-1b/vol-c85ad849" (spec.Name: "pvc-47ad28c0-7043-11e6-849c-06cdf0c19b2d") pod "2bf9e9c1-7044-11e6-849c-06cdf0c19b2d" (UID: "2bf9e9c1-7044-11e6-849c-06cdf0c19b2d") devicePath: "/dev/xvdbs"
Sep 01 13:01:26 ip-172-20-78-100 kubelet[738]: I0901 13:01:26.541447     738 reconciler.go:254] MountVolume operation started for volume "kubernetes.io/aws-ebs/aws://us-east-1b/vol-c85ad849" (spec.Name: "pvc-47ad28c0-7043-11e6-849c-06cdf0c19b2d") to pod "2bf9e9c1-7044-11e6-849c-06cdf0c19b2d" (UID: "2bf9e9c1-7044-11e6-849c-06cdf0c19b2d").  
Sep 01 13:01:26 ip-172-20-78-100 kubelet[738]: I0901 13:01:26.541507     738 operation_executor.go:649] Entering MountVolume.WaitForAttach for volume "kubernetes.io/aws-ebs/aws://us-east-1b/vol-c85ad849" (spec.Name: "pvc-47ad28c0-7043-11e6-849c-06cdf0c19b2d") pod "2bf9e9c1-7044-11e6-849c-06cdf0c19b2d" (UID: "2bf9e9c1-7044-11e6-849c-06cdf0c19b2d") DevicePath: "/dev/xvdbs"  
Sep 01 13:04:09 ip-172-20-78-100 kubelet[738]: I0901 13:04:09.541721     738 attacher.go:106] Successfully found attached AWS Volume "aws://us-east-1b/vol-c85ad849".
Sep 01 13:04:09 ip-172-20-78-100 kubelet[738]: I0901 13:04:09.541763     738 operation_executor.go:669] MountVolume.WaitForAttach succeeded for volume "kubernetes.io/aws-ebs/aws://us-east-1b/vol-c85ad849" (spec.Name: "pvc-47ad28c0-7043-11e6-849c-06cdf0c19b2d") pod "2bf9e9c1-7044-11e6-849c-06cdf0c19b2d" (UID: "2bf9e9c1-7044-11e6-849c-06cdf0c19b2d").
Sep 01 13:04:09 ip-172-20-78-100 kubelet[738]: I0901 13:04:09.577108     738 operation_executor.go:705] MountVolume.MountDevice succeeded for volume "kubernetes.io/aws-ebs/aws://us-east-1b/vol-c85ad849" (spec.Name: "pvc-47ad28c0-7043-11e6-849c-06cdf0c19b2d") pod "2bf9e9c1-7044-11e6-849c-06cdf0c19b2d" (UID: "2bf9e9c1-7044-11e6-849c-06cdf0c19b2d").
Sep 01 13:04:09 ip-172-20-78-100 kubelet[738]: I0901 13:04:09.583871     738 operation_executor.go:740] MountVolume.SetUp succeeded for volume "kubernetes.io/aws-ebs/aws://us-east-1b/vol-c85ad849" (spec.Name: "pvc-47ad28c0-7043-11e6-849c-06cdf0c19b2d") pod "2bf9e9c1-7044-11e6-849c-06cdf0c19b2d" (UID: "2bf9e9c1-7044-11e6-849c-06cdf0c19b2d").
```


But comparing those timestamps to the KCM log is interesting:

```
I0901 13:01:35.602182       6 operation_executor.go:585] DetachVolume.Detach succeeded for volume "kubernetes.io/aws-ebs/aws://us-east-1b
/vol-c85ad849" (spec.Name: "pvc-47ad28c0-7043-11e6-849c-06cdf0c19b2d") from node "ip-172-20-78-100.ec2.internal".
I0901 13:01:35.611976       6 aws.go:1263] Waiting for volume state: actual=attaching, desired=attached
I0901 13:01:35.614680       6 aws.go:1263] Waiting for volume state: actual=detaching, desired=detached
I0901 13:01:35.619151       6 reconciler.go:170] Started AttachVolume for volume "kubernetes.io/aws-ebs/aws://us-east-1b/vol-c85ad849" to
 node "ip-172-20-78-100.ec2.internal"
I0901 13:01:35.626478       6 aws.go:1263] Waiting for volume state: actual=detaching, desired=detached
I0901 13:01:35.651008       6 aws.go:1263] Waiting for volume state: actual=detaching, desired=detached
I0901 13:01:35.654322       6 pet.go:101] PetSet filetestding-19 waiting on unhealthy pet filetestding-19-1
I0901 13:01:35.654501       6 pet_set.go:317] Syncing PetSet default/filetestding-20 with 1 pets
I0901 13:01:35.702599       6 aws.go:1263] Waiting for volume state: actual=detaching, desired=detached
I0901 13:01:35.703748       6 pet_set.go:325] PetSet filetestding-20 blocked from scaling on pet filetestding-20-0
I0901 13:01:35.708149       6 aws.go:1263] Waiting for volume state: actual=detaching, desired=detached
I0901 13:01:35.718174       6 aws.go:1263] Waiting for volume state: actual=detaching, desired=detached
I0901 13:01:35.748320       6 aws.go:1263] Waiting for volume state: actual=detaching, desired=detached
I0901 13:01:35.748711       6 aws.go:1263] Waiting for volume state: actual=attaching, desired=attached
I0901 13:01:35.769322       6 aws.go:1263] Waiting for volume state: actual=detaching, desired=detached
I0901 13:01:35.777848       6 aws.go:1140] Assigned mount device bm -> volume vol-c85ad849
I0901 13:01:35.787028       6 aws.go:1263] Waiting for volume state: actual=attaching, desired=attached
I0901 13:01:35.793182       6 aws.go:1263] Waiting for volume state: actual=detaching, desired=detached
I0901 13:01:35.803669       6 pet.go:101] PetSet filetestding-20 waiting on unhealthy pet filetestding-20-0
I0901 13:01:35.809972       6 aws.go:1263] Waiting for volume state: actual=attaching, desired=attached
I0901 13:01:35.903607       6 pet.go:106] Create of filetestding-20-1 in PetSet filetestding-20 blocked by unhealthy pet filetestding-20-
0
I0901 13:01:35.903722       6 pet_set.go:317] Syncing PetSet default/filetestding-6 with 1 pets
I0901 13:01:35.953989       6 pet_set.go:325] PetSet filetestding-6 blocked from scaling on pet filetestding-6-0
I0901 13:01:36.053889       6 pet.go:101] PetSet filetestding-6 waiting on unhealthy pet filetestding-6-0
I0901 13:01:36.066749       6 aws.go:1263] Waiting for volume state: actual=detaching, desired=detached
I0901 13:01:36.153835       6 pet.go:106] Create of filetestding-6-1 in PetSet filetestding-6 blocked by unhealthy pet filetestding-6-0
I0901 13:01:36.153953       6 pet_set.go:317] Syncing PetSet default/filetestding-16 with 2 pets
I0901 13:01:36.203907       6 pet_set.go:325] PetSet filetestding-16 blocked from scaling on pet filetestding-16-1
I0901 13:01:36.256230       6 aws.go:1263] Waiting for volume state: actual=attaching, desired=attached
I0901 13:01:36.400618       6 aws.go:1385] AttachVolume request returned {
  AttachTime: 2016-09-01 13:01:36.388 +0000 UTC,
  Device: "/dev/xvdbm",
  InstanceId: "i-3e7b0cc6",
  State: "attaching",
  VolumeId: "vol-c85ad849"
}
I0901 13:01:36.403627       6 pet.go:101] PetSet filetestding-16 waiting on unhealthy pet filetestding-16-1
I0901 13:01:36.403750       6 pet_set.go:317] Syncing PetSet default/filetestding-3 with 1 pets
I0901 13:01:36.454012       6 pet_set.go:325] PetSet filetestding-3 blocked from scaling on pet filetestding-3-0
I0901 13:01:36.486960       6 aws.go:1263] Waiting for volume state: actual=attaching, desired=attached
```
 
i.e. the mounpoint bm is assigned ten seconds _after_ the mount happened on the kubelet.  The volume is still detaching at the time of the kubelet mount.

And KCM does not complete the volume mount until here:
```
I0901 13:03:52.038153       6 operation_executor.go:474] AttachVolume.Attach succeeded for volume "kubernetes.io/aws-ebs/aws://us-east-1b/vol-c85ad849" (spec.Name: "pvc-47ad28c0-7043-11e6-849c-06cdf0c19b2d") from node "ip-172-20-78-100.ec2.internal".
```

(some minutes later because of rate limits)


I'm going to need some help please @saad-ali @jingxu97  :-)

**Comment by justinsb** [comment]
(And ntp was running, so I believe these timestamps should be accurate - certainly to within a second)

**Comment by justinsb** [comment]
One more piece of information... the log message for KCM around the "problem time period" was:

```
I0901 13:00:12.854787       6 reconciler.go:115] Started DetachVolume for volume "kubernetes.io/aws-ebs/aws://us-east-1b/vol-c85ad849" fr
om node "ip-172-20-78-100.ec2.internal"
...
I0901 13:00:12.865758       6 operation_executor.go:566] Verified volume is safe to detach for volume "kubernetes.io/aws-ebs/aws://us-eas
t-1b/vol-c85ad849" (spec.Name: "pvc-47ad28c0-7043-11e6-849c-06cdf0c19b2d") from node "ip-172-20-78-100.ec2.internal".
...
I0901 13:01:35.602182       6 operation_executor.go:585] DetachVolume.Detach succeeded for volume "kubernetes.io/aws-ebs/aws://us-east-1b/vol-c85ad849" (spec.Name: "pvc-47ad28c0-7043-11e6-849c-06cdf0c19b2d") from node "ip-172-20-78-100.ec2.internal". 
...
I0901 13:01:35.619151       6 reconciler.go:170] Started AttachVolume for volume "kubernetes.io/aws-ebs/aws://us-east-1b/vol-c85ad849" to node "ip-172-20-78-100.ec2.internal" 
```

So while the kubelet was mounting the volume on bs, KCM was doing the detaching from bs and waiting for it to complete.

PS thanks for the awesome repro @mrook and @svanderbijl

**Comment by justinsb** [comment]
Could this be as simple as the pod being scheduled on the node, and the kubelet picking up a previous mount?

A few things I found suspicious (but this is just by visual scanning of the code):

* Maybe the kubelet is not observing the "volume is no longer safe to mount" here https://github.com/kubernetes/kubernetes/blob/f25d5439af327d0cc4a785850e33e93caf4d11f0/pkg/controller/volume/attachdetach/reconciler/reconciler.go#L103


* https://github.com/kubernetes/kubernetes/blob/99295356ed152663698e002494a5af00d682dc6b/pkg/controller/volume/attachdetach/statusupdater/node_status_updater.go#L61 seems to read the changes, posts the changes, and then clears the dirty flag (I think!).  But maybe if a change comes in during the post changes we will erroneously clear the dirty flag.

**Comment by jingxu97** [comment]
@justinsb Could you please share the full KCM log and kubelet log so that I can help to take alook? Thank you!

**Comment by jingxu97** [comment]
Without verifying with the logs, I think the following race condition might happen between master and kubelet

At master the following events happen
volume attach as device1 --------------------------> detach ----------------->attach on the same node as device2

At kubelet around the same time
verifyAttach by checking /dev/disk/by-id/google-[DISK_NAME] (attached as device1) -->mount from device1 ------>unmount----->verifyAttach again (the problem is that volume is still attached as device1 since detach has not yet happen or finish) -->mount from device1 again (although the correct one should device2)

**Comment by jingxu97** [comment]
After checking the log provided by @justinsb and the code path, I think the following race condition happened (similar to what I described in last comment with more details)

1. Create pod on node with volume: 
**At master** attached the volume to the node at /dev/device1, 
**At kubelet**: kubelet verify volume is attached and mounts the volume

2. Delete/cordon node, Delete pod: 
**At maste**r: reconciler started to detach the volume. At this moment, the volume is still attached so in actual_state_of_world, this volume was still stored in attachedVolumes map, but it was removed from volumesToReportAsAttached in nodesToUpdateStatusFor (this means that if kubelet checks the node now, the volume is not attached)

3. Add node back and add a new podB to the same node and podB still refers to the same volume: 
**At master**: reconciler checks volume already attached (Note: at this moment, detach is going on but not yet finished, so volume is still attached at /dev/device1) and also trigger update actual_state_world so that the volume is added back to volumesToReportAsAttached. 
**At kubelet**: reconciler verify volume is already attached through node status, and got the original device path /dev/device1, and then tries to verify the device path exist or not. 
**At master**: Now detach finishes, master updates actual_state_of_world, in turn triggers attaching volume to the same node again and this time volume is attached to /dev/device2. Also /dev/device1 is assigned to another volumeB. 
**At kubelet**: now kubelet can verify device /dev/device1 does exist and tries to mount it, but the problem is that this device1 is assigned to volume B which is wrong for new podB.

I will double check my theory by trying some tests with more logs and work on fixing this bug. Please let me know if you have comments/suggestions.

**Comment by saad-ali** [comment]
Thanks for digging into this @justinsb and @jingxu97. To summarize what is happening: attach/detach controller tries to be smart, if a volume is already attached it will not reattach the volume. If a detach is pending, however, a false signal can be given to the node that the volume is attached (even though it references the disk from the previous operation). The attach detach controller rectifies itself eventually because the detach operation completes and a new attach operation is triggered. However the kubelet at this point already mounts the old device path that then gets detached. To fix this, we will make a subtle change: 1) add volume to the list of `volumesToReportAsAttached` only if it has actually gone through a complete/real attach operation, and 2) re-add volumes to `volumesToReportAsAttached` if a detach operation fails.

**Comment by geoah** [comment]
Is there a temp fix for this until it is fixed upstream?

@svanderbijl suggested `--enable-controller-attach-detach=false` but I'm not sure what the impact of this will be. If I understand correctly it will disable the new attach/detach controller and give control of attaching/detaching storage to kubelet, but I am not sure what the downsides of this are.

> --enable-controller-attach-detach[=true]: Enables the Attach/Detach controller to manage attachment/detachment of volumes scheduled to this node, and disables kubelet from executing any attach/detach operations

**Comment by pwittrock** [comment]
@justinsb @jingxu97 @saad-ali 

I am leaving this in the 1.4 release as a blocking issue.  Please provide daily status updates to this issue or move it out of the release and into v1.4-nonblocking where it will no longer be tracked as a release blocking issue in the burndown meetings.

**Comment by saad-ali** [comment]
PR https://github.com/kubernetes/kubernetes/pull/32242 is out to fix this.

Status: PR under review

**Comment by saad-ali** [comment]
> @svanderbijl suggested --enable-controller-attach-detach=false but I'm not sure what the impact of this will be. If I understand correctly it will disable the new attach/detach controller and give control of attaching/detaching storage to kubelet, but I am not sure what the downsides of this are.

Downsides are that volumes can be stranded on a node if kubelet on that node goes down (other node can't attach volume until it is detached from the first node).

**Comment by robdaemon** [comment]
Are there plans to backport this into 1.3? This is affecting many of us that are currently running 1.3.

**Comment by justinsb** [comment]
@robdaemon yes I believe once we confirm the fix in 1.4 we will then backport it to 1.3

**Comment by matchstick** [comment]
@jingxu97 @justinsb @saad-ali @davidopp @pwittrock This fix needs to go in 1.4. David you will want to wait for this to be merged for upgrade testing. Also @jingxu97 please ask for help to ensure this gets reviewed ASAP.

**Comment by saad-ali** [comment]
The fix for this (https://github.com/kubernetes/kubernetes/pull/32242) has been LGTM'd and merged to HEAD.

@svanderbijl or @justinsb: could you please test against HEAD to sanity check and re-verify that the issue is fixed. Once we get your thumbs up, @jingxu97 will cherry-pick the PR to `release-1.3` as well.

**Comment by argusua** [comment]
Hi. We got the same issue. Volumes not usable. Data on volumes mixed after restart pods. 2 or 3 pods may use same storage at one time. 
```
/dev/xvdbc on /var/lib/kubelet/plugins/kubernetes.io/aws-ebs/mounts/vol-20a4f391 type ext4 (rw,relatime,data=ordered)
/dev/xvdbc on /var/lib/kubelet/pods/e2df197f-79a7-11e6-a1de-02b0d4bb3eab/volumes/kubernetes.io~aws-ebs/influxdb type ext4 (rw,relatime,data=ordered)
/dev/xvdbc on /var/lib/kubelet/plugins/kubernetes.io/aws-ebs/mounts/vol-a179e110 type ext4 (rw,relatime,data=ordered)
/dev/xvdbc on /var/lib/kubelet/pods/994e2015-79a6-11e6-a1de-02b0d4bb3eab/volumes/kubernetes.io~aws-ebs/sinopia-persistent-storage type ext4 (rw,relatime,data=ordered)
```

Any chance that the fix will be released in a few days?

**Comment by justinsb** [comment]
Testing against head now

**Comment by jingxu97** [comment]
Based on @justinsb 's testing, the fix (#32242) is working. It seems a little bit late to push it to release 1.4.0 now, but we will definitely try to make it in release 1.4.1 (expected early Oct). We will also work on cherry-pick the PR to release-1.3 as well. Please let me know if you have any questions/issues. Thanks!

**Comment by chrislovecnm** [comment]
@jingxu97 so this is not going into 1.4?

**Comment by jingxu97** [comment]
@chrislovecnm Sorry for the wrong information. The fix for the PR will actually go into the release 1.4.0 (it got cherry-picked to 1.4.0). Please let me know if you have any questions about it.

**Comment by chrislovecnm** [comment]
@jingxu97 👍

**Comment by pwittrock** [comment]
@justinsb I am planning on cutting another beta tonight.  Would you manually QA that this is working on Monday and then close the issue if it is?

**Comment by trizvanov** [comment]
Any ETA on when this is going to be ported to 1.3 ?
Thanks!

**Comment by matchstick** [comment]
@fabioy @timstclair Bringing this to your attention as you are both the release czars for 1.3. This is worth including in 1.3.8. Possibly even cutting a 1.3.8.

**Comment by justinsb** [comment]
This seems to be fixed in v1.4.0-beta.8

We will work on backports & further validation

**Comment by chrisleck** [comment]
Where can we track the backport to 3.x?

**Comment by jingxu97** [comment]
@pwittrock I am wondering whether we can leave it open until the fix is backported and users confirm the fix works fine?

**Comment by trizvanov** [comment]
I'd appreciate if this was still open until the backport was complete. This
issue is really hurting us in 1.3.5.

**Comment by saad-ali** [comment]
There are three PRs that address the bulk of this issue:
* PR https://github.com/kubernetes/kubernetes/pull/31090
* PR https://github.com/kubernetes/kubernetes/pull/32242
* PR https://github.com/kubernetes/kubernetes/pull/32636

As @justinsb mentioned these are all merged to head, are part of v1.4.0-beta.8 and will be part of the v1.4.0 release which is going out shortly.

We are also in the process of merging these changes back into v1.3, the relevant cherry-picks are:
* PR https://github.com/kubernetes/kubernetes/pull/33149
* PR https://github.com/kubernetes/kubernetes/pull/33142
* PR https://github.com/kubernetes/kubernetes/pull/33150

I will re-open this issue, and we can keep it open until v1.3.8 is cut/verified.

**Comment by pwittrock** [comment]
@saad-ali I think the milestone for this needs to be changed to v1.3.  I did that for you, let me know if it was the wrong this to do.

**Comment by argusua** [comment]
Looks like this not fixed in v1.4.0:
```
admin@ip-172-20-106-241:~$ mount | grep xvd
/dev/xvda1 on / type ext4 (rw,relatime,data=ordered)
/dev/xvdbb on /var/lib/kubelet/plugins/kubernetes.io/aws-ebs/mounts/vol-a179e110 type ext4 (rw,relatime,data=ordered)
/dev/xvdbb on /var/lib/kubelet/pods/3519a130-84b2-11e6-8bef-068fbb000c99/volumes/kubernetes.io~aws-ebs/sinopia-persistent-storage type ext4 (rw,relatime,data=ordered)
/dev/xvdbb on /var/lib/kubelet/plugins/kubernetes.io/aws-ebs/mounts/vol-e1580350 type ext4 (rw,relatime,data=ordered)
/dev/xvdbb on /var/lib/kubelet/pods/24f6bcd7-84b3-11e6-8bef-068fbb000c99/volumes/kubernetes.io~aws-ebs/alertmanager-storage type ext4 (rw,relatime,data=ordered)

```
```
admin@ip-172-20-106-241:~$ kubelet --version
Kubernetes v1.4.0
```
Aws console says that vol-e1580350 attached as /dev/xvdbb and vol-a179e110 as /dev/xvdba

**Comment by jingxu97** [comment]
Hi @argusua , Could you please provide more details about your cluster? Is this a GKE cluster, if so, please share us the project id, zone name, and cluster name so that we can check the master log. Could you also please share the kubelet log on the node? Thank you!

**Comment by argusua** [comment]
Hi @jingxu97 , It is AWS. Log attached.
[syslog.1.gz](https://github.com/kubernetes/kubernetes/files/497443/syslog.1.gz)

**Comment by svanderbijl** [comment]
Just ran the repro for a freshly installed 1.4.0 cluster and we can confirm the fix isn't working / included.

**Comment by svanderbijl** [comment]
Same issue with 1.3.8.

**Comment by svanderbijl** [comment]
After running a few more tests it appears the same issues exists with version 1.3.8. Was the fix applied?

**Comment by jingxu97** [comment]
@svanderbijl , the fix is not in 1.3.8. Could you please share the master kube-controller-manager log during the test if it is available? I suspect that the fix is not complete. We tested with cordon nodes approach, but not the destroy nodes.

**Comment by jingxu97** [comment]
@argusua , thank you for the log. But I still need the master kube-controller-manager log because I cannot access it. Could you please share it if it is still available?

**Comment by jingxu97** [comment]
@svanderbijl, could you also please share some details on how to shut down the cluster node? Are you just stopping (pausing) the node, and restart? Or is this a complete hard shun down (terminate) and a new node is created?

**Comment by spacepluk** [comment]
I've seen this in my cluster a few times when spot instances come back
after being terminated.

**Comment by argusua** [comment]
Hi, @jingxu97 . I've added logs of kube-controller-manager. This log from point when master node was updated to v1.4.0 and then update of minions starts. I replace nodes by marking them as Unhealthy in aws autoscaling group one by one.
[kube-controller-manager.log.2.gz](https://github.com/kubernetes/kubernetes/files/499857/kube-controller-manager.log.2.gz)

**Comment by mrook** [comment]
@jingxu97 I'll answer for @svanderbijl ; we terminated nodes and let them be replaced (by the ASG) with new nodes. This was on a cluster with  two worker nodes.

**Comment by jingxu97** [comment]
@argusua Thank you for the log. Do you remember which pod got the wrong volume attached? Thank you!

**Comment by argusua** [comment]
@jingxu97 No, I don't remember. I remember only about grafana pod, data was missing for this pod, because volume from another pod was mounted. But I don't sure now that grafana pod was on the same node.

**Comment by jingxu97** [comment]
With @justinsb help, I am able to reproduce the error and also have a fix out PR #33796 

To summarize, there are two approaches to produce this error. 
1. cordon nodes; delete pods; uncordon node; (re)create pods
2. terminate nodes; pods evicted (deleted); (re)create pods 
Although both approaches can produce the same error, they are caused by very different issues. Last time we fix issue 1 (PR #32242), but not issue 2. 

I am also working on a proposal to make the system more robust to this issue in PR #33760

Sorry for any inconvenience caused by this, and please let me know if you have any questions.

**Comment by jingxu97** [comment]
Hi @argusua @spacepluk @svanderbijl and all other AWS users, the fix #32242 is already merged. I would really appreciate if you could test against the master version (or the head of the tree) to see whether the problem is fixed or not on your cluster. 

There is still a chance of failure if you reboot the node instead of terminating the node and replace the node with new name. I am working this which requires a bigger fix in PR #33760

Please let me know if you have any questions. Thank you!

**Comment by argusua** [comment]
Thank you, @jingxu97 
Unfortunately I have a vacation and I'll come back only after 2 weeks. I hope somebody can help with testing also.

**Comment by tnine** [comment]
FYI, this issue also effects PetSets in the 1.4.0 release.  When a pod is killed or the minion running it dies, the volume does not detach to re-attach to the newly assigned minion instance.  However, the pod will start after timing out on the attach of the PV.  This seems like a bug with PetSets.  If the PV fails to mount, it should not continue.  Should this be a separate issue or included here?

**Comment by jingxu97** [comment]
@tnine, for minion dies and restart/recreate, I am trying to address the issue of volume manager fails to detach and attach in #33760. Please let me know if that PR will address your concern or not.

As for pod is killed (nodes are running file), attach/detach should work as supposed to. Please let me know if you have any issue you have experienced. Thanks!

**Comment by rdtr** [comment]
Is this issue already resolved?
I've been facing with very similar issue with v1.4.6+coreos0 for a prometheus pod.

**Comment by mikesplain** [comment]
@rdtr Not the same issue as this but may help you https://github.com/kubernetes/kubernetes/pull/36840, which should be released this coming week on the 1.4 branch.

**Comment by rdtr** [comment]
I see. Thank you for the information! I'm looking forward to the next release.

**Comment by hanikesn** [comment]
We upgraded to 1.4.7 and are still seeing delays of about 2 min, when pods change the host they're running on:

```
  FirstSeen	LastSeen	Count	From			SubobjectPath	Type		Reason		Message
  ---------	--------	-----	----			-------------	--------	------		-------
  4m		4m		1	{default-scheduler }			Normal		Scheduled	Successfully assigned grafana-11524727-6vfm2 to ip-10-128-16-121.eu-central-1.compute.internal
  4m		3m		8	{controller-manager }			Warning		FailedMount	Failed to attach volume "grafana" on node "ip-10-128-16-121.eu-central-1.compute.internal" with: Error attaching EBS volume: VolumeInUse: vol-0dacaeb7988eaecb8 is already attached to an instance
		status code: 400, request id: 
  2m		2m	1	{kubelet ip-10-128-16-121.eu-central-1.compute.internal}					Warning	FailedMount	Unable to mount volumes for pod "grafana-11524727-6vfm2_kube-system(bf839d9d-c158-11e6-8c8f-066e0d14e725)": timeout expired waiting for volumes to attach/mount for pod "grafana-11524727-6vfm2"/"kube-system". list of unattached/unmounted volumes=[data-volume]
  2m		2m	1	{kubelet ip-10-128-16-121.eu-central-1.compute.internal}					Warning	FailedSync	Error syncing pod, skipping: timeout expired waiting for volumes to attach/mount for pod "grafana-11524727-6vfm2"/"kube-system". list of unattached/unmounted volumes=[data-volume]
  41s		41s	1	{kubelet ip-10-128-16-121.eu-central-1.compute.internal}	spec.containers{grafana}	Normal	Pulling		pulling image "grafana/grafana:4.0.2"
  31s		31s	1	{kubelet ip-10-128-16-121.eu-central-1.compute.internal}	spec.containers{grafana}	Normal	Pulled		Successfully pulled image "grafana/grafana:4.0.2"
  31s		31s	1	{kubelet ip-10-128-16-121.eu-central-1.compute.internal}	spec.containers{grafana}	Normal	Created		Created container with docker id 18e5dc26b0cb; Secu
```

```
Dec 13 17:25:13 master1 kube-controller-manager[10866]: I1213 17:25:13.530908   10866 operation_executor.go:620] AttachVolume.Attach succeeded for volume "kubernetes.io/aws-ebs/vol-0dacaeb7988eaecb8" (spec.Name: "grafana") from node "ip-10-128-16-121.eu-central-1.compute.internal".
Dec 13 17:25:01 master1 kube-controller-manager[10866]: I1213 17:25:01.978733   10866 reconciler.go:202] Started AttachVolume for volume "kubernetes.io/aws-ebs/vol-0dacaeb7988eaecb8" to node "ip-10-128-16-121.eu-central-1.compute.internal"
Dec 13 17:25:01 master1 kube-controller-manager[10866]: I1213 17:25:01.935105   10866 operation_executor.go:700] DetachVolume.Detach succeeded for volume "kubernetes.io/aws-ebs/vol-0dacaeb7988eaecb8" (spec.Name: "grafana") from node "ip-10-128-16-122.eu-central-1.compute.internal".
Dec 13 17:24:51 master1 kube-controller-manager[10866]: I1213 17:24:51.317749   10866 operation_executor.go:754] Verified volume is safe to detach for volume "kubernetes.io/aws-ebs/vol-0dacaeb7988eaecb8" (spec.Name: "grafana") from node "ip-10-128-16-122.eu-central-1.compute.intern
al".
Dec 13 17:24:51 master1 kube-controller-manager[10866]: I1213 17:24:51.311087   10866 reconciler.go:167] Started DetachVolume for volume "kubernetes.io/aws-ebs/vol-0dacaeb7988eaecb8" from node "ip-10-128-16-122.eu-central-1.compute.internal"
Dec 13 17:23:47 master1 kube-controller-manager[10866]: status code: 400, request id:
Dec 13 17:23:47 master1 kube-controller-manager[10866]: I1213 17:23:47.282409   10866 event.go:217] Event(api.ObjectReference{Kind:"Pod", Namespace:"kube-system", Name:"grafana-11524727-6vfm2", UID:"bf839d9d-c158-11e6-8c8f-066e0d14e725", APIVersion:"v1", ResourceVersion:"12517206",
 FieldPath:""}): type: 'Warning' reason: 'FailedMount' Failed to attach volume "grafana" on node "ip-10-128-16-121.eu-central-1.compute.internal" with: Error attaching EBS volume: VolumeInUse: vol-0dacaeb7988eaecb8 is already attached to an instance
Dec 13 17:23:47 master1 kube-controller-manager[10866]: status code: 400, request id:
Dec 13 17:23:47 master1 kube-controller-manager[10866]: E1213 17:23:47.282382   10866 nestedpendingoperations.go:262] Operation for "\"kubernetes.io/aws-ebs/vol-0dacaeb7988eaecb8\"" failed. No retries permitted until 2016-12-13 17:24:51.282357403 +0000 UTC (durationBeforeRetry 1m4s
). Error: Failed to attach volume "grafana" on node "ip-10-128-16-121.eu-central-1.compute.internal" with: Error attaching EBS volume: VolumeInUse: vol-0dacaeb7988eaecb8 is already attached to an instance
Dec 13 17:23:47 master1 kube-controller-manager[10866]: status code: 400, request id:
Dec 13 17:23:47 master1 kube-controller-manager[10866]: E1213 17:23:47.282300   10866 attacher.go:72] Error attaching volume "vol-0dacaeb7988eaecb8": Error attaching EBS volume: VolumeInUse: vol-0dacaeb7988eaecb8 is already attached to an instance
Dec 13 17:23:46 master1 kube-controller-manager[10866]: I1213 17:23:46.844133   10866 reconciler.go:202] Started AttachVolume for volume "kubernetes.io/aws-ebs/vol-0dacaeb7988eaecb8" to node "ip-10-128-16-121.eu-central-1.compute.internal"
Dec 13 17:23:14 master1 kube-controller-manager[10866]: status code: 400, request id:
Dec 13 17:23:14 master1 kube-controller-manager[10866]: I1213 17:23:14.806039   10866 event.go:217] Event(api.ObjectReference{Kind:"Pod", Namespace:"kube-system", Name:"grafana-11524727-6vfm2", UID:"bf839d9d-c158-11e6-8c8f-066e0d14e725", APIVersion:"v1", ResourceVersion:"12517206",
 FieldPath:""}): type: 'Warning' reason: 'FailedMount' Failed to attach volume "grafana" on node "ip-10-128-16-121.eu-central-1.compute.internal" with: Error attaching EBS volume: VolumeInUse: vol-0dacaeb7988eaecb8 is already attached to an instance
Dec 13 17:23:14 master1 kube-controller-manager[10866]: status code: 400, request id:
Dec 13 17:23:14 master1 kube-controller-manager[10866]: E1213 17:23:14.805591   10866 nestedpendingoperations.go:262] Operation for "\"kubernetes.io/aws-ebs/vol-0dacaeb7988eaecb8\"" failed. No retries permitted until 2016-12-13 17:23:46.805566071 +0000 UTC (durationBeforeRetry 32s)
. Error: Failed to attach volume "grafana" on node "ip-10-128-16-121.eu-central-1.compute.internal" with: Error attaching EBS volume: VolumeInUse: vol-0dacaeb7988eaecb8 is already attached to an instance
Dec 13 17:23:14 master1 kube-controller-manager[10866]: status code: 400, request id:
Dec 13 17:23:14 master1 kube-controller-manager[10866]: E1213 17:23:14.805510   10866 attacher.go:72] Error attaching volume "vol-0dacaeb7988eaecb8": Error attaching EBS volume: VolumeInUse: vol-0dacaeb7988eaecb8 is already attached to an instance
Dec 13 17:23:14 master1 kube-controller-manager[10866]: I1213 17:23:14.238093   10866 reconciler.go:202] Started AttachVolume for volume "kubernetes.io/aws-ebs/vol-0dacaeb7988eaecb8" to node "ip-10-128-16-121.eu-central-1.compute.internal"
Dec 13 17:22:58 master1 kube-controller-manager[10866]: status code: 400, request id:
Dec 13 17:22:58 master1 kube-controller-manager[10866]: I1213 17:22:58.162915   10866 event.go:217] Event(api.ObjectReference{Kind:"Pod", Namespace:"kube-system", Name:"grafana-11524727-6vfm2", UID:"bf839d9d-c158-11e6-8c8f-066e0d14e725", APIVersion:"v1", ResourceVersion:"12517206",
 FieldPath:""}): type: 'Warning' reason: 'FailedMount' Failed to attach volume "grafana" on node "ip-10-128-16-121.eu-central-1.compute.internal" with: Error attaching EBS volume: VolumeInUse: vol-0dacaeb7988eaecb8 is already attached to an instance
Dec 13 17:22:58 master1 kube-controller-manager[10866]: status code: 400, request id:
Dec 13 17:22:58 master1 kube-controller-manager[10866]: E1213 17:22:58.162406   10866 nestedpendingoperations.go:262] Operation for "\"kubernetes.io/aws-ebs/vol-0dacaeb7988eaecb8\"" failed. No retries permitted until 2016-12-13 17:23:14.162380778 +0000 UTC (durationBeforeRetry 16s)
. Error: Failed to attach volume "grafana" on node "ip-10-128-16-121.eu-central-1.compute.internal" with: Error attaching EBS volume: VolumeInUse: vol-0dacaeb7988eaecb8 is already attached to an instance
Dec 13 17:22:58 master1 kube-controller-manager[10866]: status code: 400, request id:
Dec 13 17:22:58 master1 kube-controller-manager[10866]: E1213 17:22:58.162324   10866 attacher.go:72] Error attaching volume "vol-0dacaeb7988eaecb8": Error attaching EBS volume: VolumeInUse: vol-0dacaeb7988eaecb8 is already attached to an instance
Dec 13 17:22:57 master1 kube-controller-manager[10866]: I1213 17:22:57.694447   10866 reconciler.go:202] Started AttachVolume for volume "kubernetes.io/aws-ebs/vol-0dacaeb7988eaecb8" to node "ip-10-128-16-121.eu-central-1.compute.internal"
Dec 13 17:22:49 master1 kube-controller-manager[10866]: status code: 400, request id:
Dec 13 17:22:49 master1 kube-controller-manager[10866]: I1213 17:22:49.684951   10866 event.go:217] Event(api.ObjectReference{Kind:"Pod", Namespace:"kube-system", Name:"grafana-11524727-6vfm2", UID:"bf839d9d-c158-11e6-8c8f-066e0d14e725", APIVersion:"v1", ResourceVersion:"12517206", FieldPath:""}): type: 'Warning' reason: 'FailedMount' Failed to attach volume "grafana" on node "ip-10-128-16-121.eu-central-1.compute.internal" with: Error attaching EBS volume: VolumeInUse: vol-0dacaeb7988eaecb8 is already attached to an instance
Dec 13 17:22:49 master1 kube-controller-manager[10866]: status code: 400, request id:
Dec 13 17:22:49 master1 kube-controller-manager[10866]: E1213 17:22:49.684778   10866 nestedpendingoperations.go:262] Operation for "\"kubernetes.io/aws-ebs/vol-0dacaeb7988eaecb8\"" failed. No retries permitted until 2016-12-13 17:22:57.684751979 +0000 UTC (durationBeforeRetry 8s). Error: Failed to attach volume "grafana" on node "ip-10-128-16-121.eu-central-1.compute.internal" with: Error attaching EBS volume: VolumeInUse: vol-0dacaeb7988eaecb8 is already attached to an instance
Dec 13 17:22:49 master1 kube-controller-manager[10866]: status code: 400, request id:
Dec 13 17:22:49 master1 kube-controller-manager[10866]: E1213 17:22:49.684697   10866 attacher.go:72] Error attaching volume "vol-0dacaeb7988eaecb8": Error attaching EBS volume: VolumeInUse: vol-0dacaeb7988eaecb8 is already attached to an instance
Dec 13 17:22:49 master1 kube-controller-manager[10866]: I1213 17:22:49.255749   10866 reconciler.go:202] Started AttachVolume for volume "kubernetes.io/aws-ebs/vol-0dacaeb7988eaecb8" to node "ip-10-128-16-121.eu-central-1.compute.internal"
Dec 13 17:22:45 master1 kube-controller-manager[10866]: status code: 400, request id:
Dec 13 17:22:45 master1 kube-controller-manager[10866]: I1213 17:22:45.188568   10866 event.go:217] Event(api.ObjectReference{Kind:"Pod", Namespace:"kube-system", Name:"grafana-11524727-6vfm2", UID:"bf839d9d-c158-11e6-8c8f-066e0d14e725", APIVersion:"v1", ResourceVersion:"12517206", FieldPath:""}): type: 'Warning' reason: 'FailedMount' Failed to attach volume "grafana" on node "ip-10-128-16-121.eu-central-1.compute.internal" with: Error attaching EBS volume: VolumeInUse: vol-0dacaeb7988eaecb8 is already attached to an instance
Dec 13 17:22:45 master1 kube-controller-manager[10866]: status code: 400, request id:
Dec 13 17:22:45 master1 kube-controller-manager[10866]: E1213 17:22:45.188109   10866 nestedpendingoperations.go:262] Operation for "\"kubernetes.io/aws-ebs/vol-0dacaeb7988eaecb8\"" failed. No retries permitted until 2016-12-13 17:22:49.188084737 +0000 UTC (durationBeforeRetry 4s). Error: Failed to attach volume "grafana" on node "ip-10-128-16-121.eu-central-1.compute.internal" with: Error attaching EBS volume: VolumeInUse: vol-0dacaeb7988eaecb8 is already attached to an instance
Dec 13 17:22:45 master1 kube-controller-manager[10866]: status code: 400, request id:
Dec 13 17:22:45 master1 kube-controller-manager[10866]: E1213 17:22:45.188031   10866 attacher.go:72] Error attaching volume "vol-0dacaeb7988eaecb8": Error attaching EBS volume: VolumeInUse: vol-0dacaeb7988eaecb8 is already attached to an instance
Dec 13 17:22:44 master1 kube-controller-manager[10866]: I1213 17:22:44.744739   10866 reconciler.go:202] Started AttachVolume for volume "kubernetes.io/aws-ebs/vol-0dacaeb7988eaecb8" to node "ip-10-128-16-121.eu-central-1.compute.internal"
Dec 13 17:22:42 master1 kube-controller-manager[10866]: status code: 400, request id:
Dec 13 17:22:42 master1 kube-controller-manager[10866]: I1213 17:22:42.730993   10866 event.go:217] Event(api.ObjectReference{Kind:"Pod", Namespace:"kube-system", Name:"grafana-11524727-6vfm2", UID:"bf839d9d-c158-11e6-8c8f-066e0d14e725", APIVersion:"v1", ResourceVersion:"12517206", FieldPath:""}): type: 'Warning' reason: 'FailedMount' Failed to attach volume "grafana" on node "ip-10-128-16-121.eu-central-1.compute.internal" with: Error attaching EBS volume: VolumeInUse: vol-0dacaeb7988eaecb8 is already attached to an instance
Dec 13 17:22:42 master1 kube-controller-manager[10866]: status code: 400, request id:
Dec 13 17:22:42 master1 kube-controller-manager[10866]: E1213 17:22:42.730470   10866 nestedpendingoperations.go:262] Operation for "\"kubernetes.io/aws-ebs/vol-0dacaeb7988eaecb8\"" failed. No retries permitted until 2016-12-13 17:22:44.730445015 +0000 UTC (durationBeforeRetry 2s). Error: Failed to attach volume "grafana" on node "ip-10-128-16-121.eu-central-1.compute.internal" with: Error attaching EBS volume: VolumeInUse: vol-0dacaeb7988eaecb8 is already attached to an instance
Dec 13 17:22:42 master1 kube-controller-manager[10866]: status code: 400, request id:
Dec 13 17:22:42 master1 kube-controller-manager[10866]: E1213 17:22:42.730392   10866 attacher.go:72] Error attaching volume "vol-0dacaeb7988eaecb8": Error attaching EBS volume: VolumeInUse: vol-0dacaeb7988eaecb8 is already attached to an instance
Dec 13 17:22:42 master1 kube-controller-manager[10866]: I1213 17:22:42.238802   10866 reconciler.go:202] Started AttachVolume for volume "kubernetes.io/aws-ebs/vol-0dacaeb7988eaecb8" to node "ip-10-128-16-121.eu-central-1.compute.internal"
Dec 13 17:22:41 master1 kube-controller-manager[10866]: status code: 400, request id:
Dec 13 17:22:41 master1 kube-controller-manager[10866]: I1213 17:22:41.196712   10866 event.go:217] Event(api.ObjectReference{Kind:"Pod", Namespace:"kube-system", Name:"grafana-11524727-6vfm2", UID:"bf839d9d-c158-11e6-8c8f-066e0d14e725", APIVersion:"v1", ResourceVersion:"12517206", FieldPath:""}): type: 'Warning' reason: 'FailedMount' Failed to attach volume "grafana" on node "ip-10-128-16-121.eu-central-1.compute.internal" with: Error attaching EBS volume: VolumeInUse: vol-0dacaeb7988eaecb8 is already attached to an instance
Dec 13 17:22:41 master1 kube-controller-manager[10866]: status code: 400, request id:
Dec 13 17:22:41 master1 kube-controller-manager[10866]: E1213 17:22:41.196686   10866 nestedpendingoperations.go:262] Operation for "\"kubernetes.io/aws-ebs/vol-0dacaeb7988eaecb8\"" failed. No retries permitted until 2016-12-13 17:22:42.196661757 +0000 UTC (durationBeforeRetry 1s). Error: Failed to attach volume "grafana" on node "ip-10-128-16-121.eu-central-1.compute.internal" with: Error attaching EBS volume: VolumeInUse: vol-0dacaeb7988eaecb8 is already attached to an instance
Dec 13 17:22:41 master1 kube-controller-manager[10866]: status code: 400, request id:
Dec 13 17:22:41 master1 kube-controller-manager[10866]: E1213 17:22:41.196609   10866 attacher.go:72] Error attaching volume "vol-0dacaeb7988eaecb8": Error attaching EBS volume: VolumeInUse: vol-0dacaeb7988eaecb8 is already attached to an instance
Dec 13 17:22:40 master1 kube-controller-manager[10866]: I1213 17:22:40.735292   10866 reconciler.go:202] Started AttachVolume for volume "kubernetes.io/aws-ebs/vol-0dacaeb7988eaecb8" to node "ip-10-128-16-121.eu-central-1.compute.internal"
Dec 13 17:22:40 master1 kube-controller-manager[10866]: status code: 400, request id:
Dec 13 17:22:40 master1 kube-controller-manager[10866]: I1213 17:22:40.147534   10866 event.go:217] Event(api.ObjectReference{Kind:"Pod", Namespace:"kube-system", Name:"grafana-11524727-6vfm2", UID:"bf839d9d-c158-11e6-8c8f-066e0d14e725", APIVersion:"v1", ResourceVersion:"12517206", FieldPath:""}): type: 'Warning' reason: 'FailedMount' Failed to attach volume "grafana" on node "ip-10-128-16-121.eu-central-1.compute.internal" with: Error attaching EBS volume: VolumeInUse: vol-0dacaeb7988eaecb8 is already attached to an instance
Dec 13 17:22:40 master1 kube-controller-manager[10866]: status code: 400, request id:
Dec 13 17:22:40 master1 kube-controller-manager[10866]: E1213 17:22:40.147477   10866 nestedpendingoperations.go:262] Operation for "\"kubernetes.io/aws-ebs/vol-0dacaeb7988eaecb8\"" failed. No retries permitted until 2016-12-13 17:22:40.647450881 +0000 UTC (durationBeforeRetry 500ms). Error: Failed to attach volume "grafana" on node "ip-10-128-16-121.eu-central-1.compute.internal" with: Error attaching EBS volume: VolumeInUse: vol-0dacaeb7988eaecb8 is already attached to an instance
Dec 13 17:22:40 master1 kube-controller-manager[10866]: status code: 400, request id:
Dec 13 17:22:40 master1 kube-controller-manager[10866]: E1213 17:22:40.147394   10866 attacher.go:72] Error attaching volume "vol-0dacaeb7988eaecb8": Error attaching EBS volume: VolumeInUse: vol-0dacaeb7988eaecb8 is already attached to an instance
Dec 13 17:22:39 master1 kube-controller-manager[10866]: I1213 17:22:39.632706   10866 reconciler.go:202] Started AttachVolume for volume "kubernetes.io/aws-ebs/vol-0dacaeb7988eaecb8" to node "ip-10-128-16-121.eu-central-1.compute.internal"
Dec 13 17:22:39 master1 kube-controller-manager[10866]: I1213 17:22:39.588878   10866 event.go:217] Event(api.ObjectReference{Kind:"ReplicaSet", Namespace:"kube-system", Name:"grafana-11524727", UID:"57a37ad9-be23-11e6-b98a-069d86b508b9", APIVersion:"extensions", ResourceVersion:"12081091", FieldPath:""}): type: 'Normal' reason: 'SuccessfulCreate' Created pod: grafana-11524727-6vfm2
```

**Comment by chrislovecnm** [comment]
@justinsb we have recreated this twice in a row.  We are going to get logs for you.

**Comment by jingxu97** [comment]
The original issue on this post about "Wrong AWS volume can be mounted" is solved already. The other issue mentioned related to detach delay is addressed in other issues https://github.com/kubernetes/kubernetes/issues/52735, https://github.com/kubernetes/kubernetes/issues/53059 . Close this one.
